sql/core/src/main/scala/org/apache/spark/sql/classic/Dataset.scala

```diff
@@ -1544,6 +1544,35 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * Repartitions the Dataset using the specified partition ID expression.
+   *
+   * The number of partitions will be configured by `spark.sql.shuffle.partitions`.
+   *
+   * @param partitionIdExpr the expression to be used as the partition ID. Must be an integral type.
+   *
+   * @group typedrel
+   * @since 4.1.0
+   */
+  def repartitionById(partitionIdExpr: Column): Dataset[T] = {
```

Comment:
I feel it's risky to provide a default numPartitions. Can we always ask users to specify numPartitions?